### PR TITLE
[10.11] Enable HDR/DV Device Profiles Conditions

### PR DESCRIPTION
**Pull request by JPKribs**
## Summary

Resolves: https://github.com/jellyfin/Swiftfin/issues/1852

Adds `VideoRangeType`s to the Device Profiles for the Swiftin & Native Players. I don't have a ton of DV for testing but what I do have tests appropriately.

VLCKit:
- `.sdr`
- `.hdr10`
- `.hdr10Plus`
- `.doviWithSDR`
- `.doviWithHDR10`
- `.doviWithHDR10Plus`

AVKit:          
- `.sdr`
- `.hdr10`
- `.hdr10Plus`
- `.dovi`
- `.doviWithSDR`
- `.doviWithHDR10`
- `.doviWithHDR10Plus`

If any of these are incorrect, please let me know. I've been able to test all of these I believe. Some of what is tricky is there are some VLCKit are invalid formats but VLC is able to handle tone mapping inside the player. Primarily, it looks like DV without a fallback is where issues originate.

In my searching for AVKit, I believe these are all correct but If anyone is more familiar with these types let me know. I tend to avoid these as I don't have a TV for it and, even if I did, I cannot personally tell the difference.

---

I also updated the `ProfileCondition.value` to be a closure for ease of input and mapping. This required me to create a `PipeStringBuilder` which is just the `ComaStringBuilder` but with Pipes. If there is a good way to combine those, that might be best.

---

### Before

https://github.com/user-attachments/assets/fb3f2f70-afb7-4a37-8dfc-f3f605d69906

### After

https://github.com/user-attachments/assets/ae1bcf47-ca95-41f3-8838-b99a51171127

### CODE REVIEW

**Comment by JPKribs** [comment]
I did some cleanup for `VideoProfiles` as well so we could use them in the same closure as other `ProfileCondition`s. This isn't a requirement for this PR. Feel free to let me know if we want to revert this.

---

Also added `.eac3` to the Native Player. I'm not sure where that got dropped but we have it documented as a supported audio type and it's one of the few advertised AVKit codecs. I don't think this resolves https://github.com/jellyfin/Swiftfin/issues/1848 but I haven't been able to recreate that issue so this could be related?

File: Shared/Objects/VideoPlayerType/VideoPlayerType+Shared.swift [hunk 1]
```diff
@@ -11,76 +11,81 @@ import JellyfinAPI
...
-                )
+    // MARK: - Shared Codec Profile Conditions
 
-                ProfileCondition(
-                    condition: .lessThanEqual,
-                    isRequired: false,
-                    property: .videoLevel,
-                    value: "175"
-                )
+    static var _h264BaseConditions: [ProfileCondition] {
```
Comment (LePips) [inline Shared/Objects/VideoPlayerType/VideoPlayerType+Shared.swift:29]: Attach these with `@ArrayBuilder<ProfileCondition>`

File: Shared/Objects/VideoPlayerType/VideoPlayerType+Swiftfin.swift [hunk 2]
```diff
@@ -137,4 +137,124 @@ extension VideoPlayerType {
             SubtitleFormat.xsub
         }
     }
+
+    // MARK: - Codec Profiles
+
+    @ArrayBuilder<CodecProfile>
+    static var _swiftfinCodecProfiles: [CodecProfile] {
+        CodecProfile(
```
Comment (LePips) [inline Shared/Objects/VideoPlayerType/VideoPlayerType+Swiftfin.swift:145]: Use the init as defined in `CodecProfile` with the `ArrayBuilder` parameters for the profiles in this file.

**Comment by JPKribs** [comment]
> I trust that the values are correct as I also won't be able to test them myself anytime soon, just some housekeeping.

While I am working on this, do we want a toggle to force HDR transcoding? I know some other clients have this. This would just take these and change this to `.sdr` as `isRequired: true`

**Comment by JPKribs** [comment]
Revisions are made and tested. On Server, in testing on my last commit, you can see this is reporting and tone mapping / transcoding appropriately. I've included a screenshot of Swiftfin player on Main.

Screenshot 2025-12-18 at 18 30 19

### Native Player - No Transcode

NativePlayer - No Transcode

### Swiftfin - This PR

Swiftfin - Transcode

### Swiftfin - Pre-PR

Swiftfin - Pre-PR

### Notice/Question

VLC actually handles HDR10 and the documentation should reflect this now. VLC is also able to locally handle HDR10+ by treating it as HDR10. This isn't a direct usage of HDR10+ so I dunno if this is better to transcode or handle as HDR10. Transcoding probably handles this more "color accurate" but IMO, I'd rather just take the HDR10 version of HDR10+ than force the transcoding?

I've updated our documentation for this. For now, this PR is allowing HDR10+ since VLC will just ignore the dynamic metadata and handle this as HDR10.

I don't really use HDR10+ or DV so like... I'm not positive what is optimal here. I've taken the "Prioritize Direct Play" route and, if someone has stronger opinions, I am happy to adjust that or, if it makes sense, adding a like HDR toggle or something.

Screenshots I went for something with skin tones since I think that is the most obvious issues for DV. Hopefully, that is a good example/test.

**Comment by holow29** [comment]
IMO it doesn't make sense to transcode HDR10+ if HDR10 is supported. Just like DoVi with HDR10 fallback is better to fallback to HDR10 on devices that don't support DoVi. Maybe in the future, there could be an option for it, but I don't think tonemapping server-side is going to produce a better image currently anyway - I actually think the HDR10 base layer will be more color-accurate.

**Comment by JPKribs** [comment]
> IMO it doesn't make sense to transcode HDR10+ if HDR10 is supported. Just like DoVi with HDR10 fallback is better to fallback to HDR10 on devices that don't support DoVi. Maybe in the future, there could be an option for it, but I don't think tonemapping server-side is going to produce a better image currently anyway - I actually think the HDR10 base layer will be more color-accurate.

Good to know! I assumed as much but I appreciate a more qualified opinion haha

**Review by LePips — APPROVED** [review]

**Comment by DrHurt** [comment]
> IMO it doesn't make sense to transcode HDR10+ if HDR10 is supported. Just like DoVi with HDR10 fallback is better to fallback to HDR10 on devices that don't support DoVi. Maybe in the future, there could be an option for it, but I don't think tonemapping server-side is going to produce a better image currently anyway - I actually think the HDR10 base layer will be more color-accurate.

+1

in my experience avplayer does an excellent job automatically falling back or tone mapping as required (as long as HLS manifest is compliant with Apple requirements which JellyFin servers seem to do very well anyway so no problem there)

**Comment by DrHurt** [comment]
@JPKribs @LePips 

AVKit also supports Dolby Vision Profile 8.4:
`doviWithHLG`

Direct play profiles seems to be missing:
mp4 + `VideoCodec.h264`
m4v + `VideoCodec.hevc`

May I suggest those also get added to the device profile.

Thanks

**Comment by JPKribs** [comment]
> AVKit also supports Dolby Vision Profile 8.4:
> doviWithHLG

Only for iPhone 12 and above:

https://professionalsupport.dolby.com/s/article/A-Guide-to-Dolby-Vision-on-iOS-From-Capture-to-Playback?language=en_US

Swiftfin supports iOS 16 which includes devices that do not have the appropriate SoC for this:

https://support.apple.com/en-us/103267

This was originally left out since I don't have a way to test this personally and this would create issues for a lot of supported devices. It might be worth seeing if we can add/remove based on SoC but for now I'm not personally comfortable adding what I cannot test.

**Comment by JPKribs** [comment]
Actually, there is a way to pull supported HDR from AVPlayer that would be better for us to use. I can make a PR for that.

---

Nevermind. There is an `AVPlayer.availableHDRModes` and `AVPlayer.eligibleForHDRPlayback` but these are purely for whether the device is **connected to an HDR display** but does not tell us if the device is capable of it.

**Comment by DrHurt** [comment]
I’m just brain storming here.

iPhone 11 support ends this year. It’s a very old device now. 

All iPhones 12 to 17 record in DV 8.4 **by default** so it’s an important profile to support for playing home videos.

Also, it looks like AVPlayer supports pure HLG as well which I can’t see in Swiftfin’s profile.

**Comment by JPKribs** [comment]
You are welcome to create a PR with the changes to discuss this further. I cannot personally test these formats so I will not be making a PR for an item I cannot test.

I spent the weekend looking at and my only real takeaway is the profile for MP4 and M4V should be merged. It also might be prudent to have a toggle for tvOS to transcode HDR content if `AVPlayer.eligibleForHDRPlayback` is false.

**Comment by DrHurt** [comment]
> You are welcome to create a PR with the changes to discuss this further. I cannot personally test these formats so I will not be making a PR for an item I cannot test.
> 
> I spent the weekend looking at and my only real takeaway is the profile for MP4 and M4V should be merged. It also might be prudent to have a toggle for tvOS to transcode HDR content if `AVPlayer.eligibleForHDRPlayback` is false.

I don’t currently have the setup to provide PRs so I’m just reporting as an end user.

Sounds sensible to merge mp4 and m4v though as they’re practically the same.

AVPlayer is capable of client side tone mapping when the display isn’t HDR compatible (but that depends on the HLS manifest). 

To get AVPlayer to tone map HDR/DV to SDR, you need to pass on the VOD playlist (the one with mp4 segments list) to AVPlayer not the master.m3u8 (tested on macOS not tvOS)

In my humble opinion, that is the more colour accurate and resource efficient way to do it.

**Comment by JPKribs** [comment]
> I don’t currently have the setup to provide PRs so I’m just reporting as an end user.

Gotcha. I can try and look into something today. 

~It's been on my list for a while to just **make** some test files for these types of things. Just something with various audio tracks and some variations with DV, HDR10+ etc. that are very basic like "this color should be orange".~

~Do you happen to know about anything like this for testing DV/HDR?~

Edit: Turns out some folks on the Server team just have these ready to go! I can test this more fully today and get a PR out there.

**Comment by DrHurt** [comment]
> > I don’t currently have the setup to provide PRs so I’m just reporting as an end user.
> 
> Gotcha. I can try and look into something today.
> 
> ~It's been on my list for a while to just **make** some test files for these types of things. Just something with various audio tracks and some variations with DV, HDR10+ etc. that are very basic like "this color should be orange".~
> 
> ~Do you happen to know about anything like this for testing DV/HDR?~
> 
> Edit: Turns out some folks on the Server team just have these ready to go! I can test this more fully today and get a PR out there.

some samples can be found here.
https://fra1.mirror.jellyfin.org/test-videos/

Bear in mind that AVPlayer behaves differently when playing HDR/DV mp4 (direct play) vs mkv -> HLS (remux & direct stream from server).

I suggest testing with DoVi P5 in MKV (because it has no HDR10 or SDR fall back so turns green/purple if not implemented & decoded correctly). Or just fails to play altogether.

HDR/DV mp4 are automatically tone mapped by AVPlayer if display is SDR (via direct play) 

HLS will NOT tone map on an SDR display if you try to play the server-provided master.m3u8 (because it tries to force mode switch to HDR which doesn’t exist so I get a black screen)

If you pass the VOD playlist (instead of master.m3u8), AVPlayer will tone map to SDR (regardless of display being HDR or SDR).

in short, you could use AVPlayer.eligibleForHDRPlayback to determine whether to play server provided master.m3u8 on HDR capable displays or pass on the VOD playlist instead on unsupported displays.

**Comment by DrHurt** [comment]
That said, I don’t know how reliable AVPlayer.eligibleForHDRPlayback is for determining whether TV is HDR ready (meaning hardware is capable and tvOS is set to HDR/DV or SDR with match range enabled).

Someone will need to test that.

ps. Upstream server team are aware of this master vs VOD playlist issue affecting Apple tone mapping and are thinking of a reliable way to implement it server side.

**Comment by JPKribs** [comment]
> https://fra1.mirror.jellyfin.org/test-videos/

I just found that as well! That helps a ton. Do you think this accurately captures our edge cases? Is there any value in 4K as well? I only bought this mac to work on Swiftfin so I cheaped out on storage so I'm shooting to keep this as small as possible haha

Screenshot 2026-01-05 at 14 14 37

**Comment by DrHurt** [comment]
no need for 4k

Only thing missing here would be DV AV1 aka P10 (which is supported by A17 (pro) / M3 & newer). 
Current Apple TV chipset is too old for AV1 but next ATV will almost certainly support it.

test file: https://mega.nz/file/aPBiACoT#A3BqP2CXQrTguj-K82qWO4jrTft1w1UDFn2oEfES-FE (untested by me - but needs audio transcoding) 

I would keep testing to mkv files as they are the tricky ones that require server streaming via HLS (without transcoding if everything is setup correctly).

**Comment by DrHurt** [comment]
You can use Subler (free OSS macOS app) to remux that AV1 DV P10 mkv to mp4 (and it can also convert the DTS audio to AAC).

The server should however stream the video directly & only transcode the audio if everything is setup correctly (will only work with A17 pro & regular A18 and newer) if you play the mkv.

**Comment by JPKribs** [comment]
> will only work with A17 pro & regular A18 and newer

Hm. That gets me back into trouble. I only have an iPhone 13 Pro so I can see that the color is incorrect but I am not able to test if we are able to `DirectPlay` on a newer device. The good news is I found a way to get the [SoC from the Metal APIs](https://developer.apple.com/documentation/metal/detecting-gpu-features-and-metal-software-versions) but now I'm realizing I can only really cater the resolves to the hardware I own haha:

Screenshot 2026-01-05 at 15 43 16

**Comment by DrHurt** [comment]
Swiftfin AVPlayer profile already seems to support AV1 including HDR/DV.

mp4 should direct play, and mkv should direct stream without transcoding on newer devices.

on older devices, it would need to transcode or fallback to vlc

———————————-

I think you should also look into how AVPlayer.eligibleForHDRPlayback behaves on tvOS. Is impacted by setting ATV output to SDR with match content on / off, HDR & DV?

I say that because setting the ATV to output SDR with match content off breaks HDR via HLS (master.m3u8) in my testing. File completely fails to play. MP4 files tone map & play though.

skipping master.m3u8 and playing the vod playlist directly fixes HDR via HLS on SDR mode by tone mapping just like direct playing the mp4.

this isn’t the case with iPhones as their internal display identifies as HDR since at least iPhone 12.

I did my testing a while back so might be worth rechecking on tvOS 26 to see if this behaviour persists.

**Comment by JPKribs** [comment]
Does this look right?

```swift
    var supportsHEVCDecode: Bool {
        (isAppleSilicon && self >= .apple3) || (isIntelMac && self >= .mac2)
    }

    var supportsAV1Decode: Bool {
        isAppleSilicon && self >= .apple9
    }

    var supportsHDR10Decode: Bool {
        isAppleSilicon && self >= .apple4
    }

    var supportsHLGDecode: Bool {
        isAppleSilicon && self >= .apple4
    }

    var supportsDolbyVisionDecode: Bool {
        isAppleSilicon && self >= .apple5
    }
```

For reference:

GPU Family | Chips
-- | --
apple1 | A7, A8
apple2 | A8
apple3 | A9, A10
apple4 | A11
apple5 | A12
apple6 | A13
apple7 | A14, M1
apple8 | A15, A16, M2
apple9 | A17 Pro, M3
apple10 | A18, A18 Pro, A19, A19 Pro, M4, M5
mac1 & mac2 | Intel Macs

**Comment by JPKribs** [comment]
> I think you should also look into how AVPlayer.eligibleForHDRPlayback behaves on tvOS. Is impacted by setting ATV output to SDR with match content on / off, HDR & DV?

https://developer.apple.com/documentation/avfoundation/avplayer/eligibleforhdrplayback

I will make sure to test this prior to any PR but this looks like this is purely for a display match. I don't know if that is the setting on the AppleTV but the simulator is appropriately returning false for my non-HDR mac. I'll do more digging/testing into this but IMO this looks promising:

> Indicates whether HDR content can be played to an appropriate display. This property is YES if an HDR display is available and the device is capable of playing HDR content from an appropriate AVAsset, NO otherwise.

I think the goal would be a `serverMappingOnIncompatible` setting or something to enable transcoding from the server if desirable over on device. Up for debate I'll need to figure out what this all looks like on the testing side. I'm still bogged down in Metal APIs so I haven't had a chance to do anything except see if we are returning the device name/family.

---

PR is still a work in progress I need to test more but:

https://github.com/jellyfin/Swiftfin/pull/1891

**Comment by JPKribs** [comment]
Testing on iPhone 13 Pro, 8.4 always fails on Native Player. iOS 26.2, I do not believe 8.4 / HLG is supported on iOS.
